field/src/extension/binomial_extension.rs

```diff
@@ -0,0 +1,368 @@
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Sub<F> for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn sub(self, rhs: F) -> Self {
+        let mut res = self.0;
+        res[0] -= rhs;
+        Self(res)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> SubAssign for BinomialExtensionField<F, D> {
+    #[inline]
+    fn sub_assign(&mut self, rhs: Self) {
+        *self = *self - rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> SubAssign<F> for BinomialExtensionField<F, D> {
+    #[inline]
+    fn sub_assign(&mut self, rhs: F) {
+        *self = *self - rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Mul for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn mul(self, rhs: Self) -> Self {
+        let Self(a) = self;
+        let Self(b) = rhs;
+
+        let mut res = [F::ZERO; D];
+
+        for i in 0..D {
+            for j in 0..D {
+                if i + j >= D {
+                    res[i + j - D] += F::W * a[i] * b[j];
+                } else {
+                    res[i + j] += a[i] * b[j];
+                }
+            }
+        }
+
+        Self(res)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Mul<F> for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[inline]
+    fn mul(self, rhs: F) -> Self {
+        Self(
+            self.0
+                .into_iter()
+                .map(|x| x * rhs)
+                .collect::<Vec<F>>()
+                .try_into()
+                .unwrap(),
+        )
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Product for BinomialExtensionField<F, D> {
+    fn product<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.fold(Self::ONE, |acc, x| acc * x)
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> Div for BinomialExtensionField<F, D> {
+    type Output = Self;
+
+    #[allow(clippy::suspicious_arithmetic_impl)]
+    fn div(self, rhs: Self) -> Self::Output {
+        self * rhs.inverse()
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> DivAssign for BinomialExtensionField<F, D> {
+    fn div_assign(&mut self, rhs: Self) {
+        *self = *self / rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> MulAssign for BinomialExtensionField<F, D> {
+    #[inline]
+    fn mul_assign(&mut self, rhs: Self) {
+        *self = *self * rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> MulAssign<F> for BinomialExtensionField<F, D> {
+    fn mul_assign(&mut self, rhs: F) {
+        *self = *self * rhs;
+    }
+}
+
+impl<F: BinomiallyExtendable<D>, const D: usize> AbstractExtensionField<F>
+    for BinomialExtensionField<F, D>
+{
+    const D: usize = F::D;
+
+    fn from_base(b: F) -> Self {
+        Self(field_to_array::<F, D>(b))
+    }
+
+    fn from_base_slice(bs: &[F]) -> Self {
+        assert_eq!(bs.len(), D);
+        Self(bs.try_into().unwrap())
```

Comment:
I'm not sure the `assert_eq!` adds anything not already provided by the `.unwrap()`.
```suggestion
        Self(bs.try_into().unwrap())
```
If you feel it's worth adding an explanatory message on failure, then I think using `.expect(...)` would be better:
```suggestion
        Self(bs.try_into().expect("slice was not expected length"))
```

Reply:
updated, thx.